### ISSUE TITLE
provide a builtin way to show an inventory without allowing to interact with it

### ISSUE THREAD

**Issue by AFCMS** [minetest/minetest#11334]
## Problem
- I want to create a tool able to see (not modify!) the content of a chest even in protected areas
- chest/machine have already `allow_move` etc callback used internally, so I cant override them with ease

There is two solutions:

* create an ugly form like that
```
local stack = inv:get_stack("main", i)
form = form.."item_image["..(i+i*0.25)-1 ..",0.25;1,1;"..stack:get_name().." "..stack:get_count().."]"
form = form.."tooltip["..(i+i*0.25)-1 ..",0.25;1,1;"..stack:get_description().."]"
```
this is bad: huge form to send to client, styling not like real `list[]`, etc

* use detached temp inventories

```
local function enpty_func() return 0 end

local callbacks = {
	allow_move = enpty_func,
	allow_put = enpty_func,
	allow_take = enpty_func,
}

local tmp_inventories = {}

minetest.register_on_joinplayer(function(player)
	local playername = player:get_player_name()
	tmp_inventories[playername] = minetest.create_detached_inventory("pala_looting:chest_explorer_"..playername,
		callbacks, playername)
end)

minetest.register_on_leaveplayer(function(player)
	local playername = player:get_player_name()
	minetest.remove_detached_inventory("pala_looting:chest_explorer_"..playername)
	tmp_inventories[playername] = nil
end)
```
```
if placer:is_player() and placer:get_player_control().sneak and pointed_thing.type == "node" then
  local playername = placer:get_player_name()
  local meta = minetest.get_meta(pointed_thing.under)
  local inv = meta:get_inventory()
  local size = inv:get_size("main")
  local list = inv:get_list("main")
  local tmp_inv = tmp_inventories[playername]

  if not tmp_inv then return end
  
  tmp_inv:set_size("main", size)
  tmp_inv:set_list("main", list)
  
  local width = size / 9
  
  local form = table.concat({
      "formspec_version[4]",
      "size[11.50,"..(width+width*0.25)+0.25 .."]",
      "list[detached:pala_looting:chest_explorer_"..playername..";main;0.25,0.25;9,"..width..";]"
  })
```

this code is bad: complex managing of temp inventories, update isn't in real time, and you can still try to pick an item in the inventory as the callbacks are server side (it should be serverside to avoid cheating AND clientside)

## Solutions

add an optional param to `list[]` to disallow interacting with the inventory

## Alternatives

add a new form element

**Comment by AFCMS** [comment]
@SmallJoker Its not exactly read-only inventory lists, but disallowing any interaction with the inventory inside a formspec

**Comment by SmallJoker** [comment]
@AFCMS Just like read-only `textarea[]` elements. You can select them, copy text, but not modify. Same goes for your idea: show items, but disallow any interaction.

**Comment by AFCMS** [comment]
ok!

**Comment by Sokomine** [comment]
Perhaps a new formspec element might not be the worst solution here. Sometimes (trading, machine setup) it is useful to be able to show a "fake" inventory to the player that looks like an inventory but doesn't really have to *be* an inventory that exists anywhere (i.e. a trade offer).
The drawback might be that something that looks like an inventory but isn't might confuse players. A diffrent style might help here: I.e. make the background of those fake inventory slots brighter.
Tooltips would still be useful.

**Comment by hlqkj** [comment]
I have just one beg here: please style it so that it clearly and immediately appears a non-interacting inventory, and not something confusing inexperienced players.

**Comment by AFCMS** [comment]
yes, it would be useful to separate read only and normal inventory list styling

**Comment by velartrill** [comment]
maybe a better direction for this issue would be an expansion of the `allow_*` API. the issue i see has nothing to do with formspecs themselves (formspecs are best thought of as the "HTML" to the backend's lua -- they're just text that gets sent to the client), but the way they're tied to nodes and the non-node-centric formspec API is pretty awkward and painful to use even insecurely. one possibility would be allowing itemstacks to have their own formspecs and associated callbacks; this would add a lot of neat options for modding (i've run into similar difficulties myself writing a mod)

@Sokomine best solve that with styles, not new elements. nothing about formspecs can be used for secure server-side checks (it's the equivalent of trying to prevent a user from uploading a file without logging in only by putting "disabled" in the HTML tag). just making a new element would also likely trip up mod devs who don't realize they need backend logic

@AFCMS in your particular case, i believe you can just cheat and use a normal inventory list, relying on the node's `allow_*` callbacks to implement the necessary protection logic and prevent the user from using them to move anything around. if they don't implement that logic, the user can steal from the chest anyway, either by way of l33t hax, or simple by right-clicking on the chest normally, regardless of whether they have the special tool.

**Comment by AFCMS** [comment]
> @AFCMS in your particular case, i believe you can just cheat and use a normal inventory list, relying on the node's `allow_*` callbacks to implement the necessary protection logic and prevent the user from using them to move anything around.

This don't work so easily:
- nodes callbacks can be already defined , so it would require making temp inventories with the right callbacks functions (which is ugly)
- using callbacks still generate unnecessary communication between the client and the server.

> if they don't implement that logic, the user can steal from the chest anyway, either by way of l33t hax, or simple by right-clicking on the chest normally, regardless of whether they have the special tool.

the check should be first client side (to improve performances and reduce network consumption) and then server side (to avoid cheaters)

**Comment by velartrill** [comment]
@AFCMS you misunderstand me; i'm sorry if i was unclear. the necessary checks are *already in place,* by definition. you don't need to write any. opening a window with an inventory list does not give the client any powers he would not have had already. the only downside is that the client "pretends" the user can modify the contents of the list, meaning he can drag things around with the mouse, but dropping them will not change anything, UI-wise or on the server.

i agree it's useful to have a way of displaying inventory lists that are not "interactable," this is why i think the engine should add a styling mechanism that allows making all widgets (not just `list[]`s) readonly. i think using the styling system for this rather than a separate element is ideal because that way it's less likely to trick devs into thinking that it gives them any extra security. currently no such readonly list mechanism exists, but there is a workaround:

you can iterate over the inventories of an object with a for loop, generating a formspec string based on each item using the `item_image[]` and `tooltip[]` elements, then sending that formspec to the user. this is suboptimal because of the large amount of text that needs to be sent and the fact that it won't live-update, but it's the best that can be done without engine changes or abusing `list[]`s. it's been a whlie since i've written any MT lua but here's some (very rough, not tested even for syntax errors) pseudocode showing the basic principle:

```lua
function form()
	return {
		src = "";
		width = 8;
		pad = 0;
		curs = {x = 0, y = 0, maxh = 0};
		nl = function(self, h)
			h = h or 0
			self.curs.x = 0
			self.curs.y = math.max(h, self.curs.y + self.curs.maxh)
			self.curs.maxh = 0
		end;
		atttach = function(self, elt, x, y, w, h, ...)
			local content = ''
			if self.width - self.curs.x < w then self:nl() end
			for _, v in pairs{...} do
				content = content .. ';' .. minetest.formspec_escape(tostring(v))
			end
			self.src = self.src .. string.format('%s[%f,%f;%f,%f%s]', elt, x,y, w,h, content)
			if h > self.curs.maxh then self.curs.maxh = h end
		end;
		add = function(self, elt, w, h, ...)
			local ax, ay = self.curs.x, self.curs.y
			self:attach(elt, ax,ay, w,h, ...)
			self.curs.x = self.curs.x + w + self.pad
			if self.curs.x > self.width then self:nl() end
			return ax, ay
		end;
		render = function()
			return string.format("size[%f,%f]%s", self.width, self.curs.y + self.pad + self.curs.maxh, self.src)
		end;
	}
end

function on_rightclick(pos, node, clicker)
	if not pos then return end
	local m = minetest.get_meta(pos)
	local inv = m:get_inventory()
	local maxwidth = 8

	local function item_view(ui, which)
		local desc = which:get_description()
		if which:get_count() > 1 then
			desc = desc .. string.format('(%u)', which:get_count())
		end
		
		local ix, iy = ui:add('item_image', 1,1, which:get_name())
		ui:attach('tooltip', ix, iy, 1,1, desc)
	end

	local function scan_inventory(ui, lst)
		ui:add('hypertext', ui.width, 1, '', lst)
		ui:nl()
		local sz = inv:get_size(lst)
		for i=1,sz do
			item_view(ui, inv:get_stack(lst, i))
		end
		ui:nl(0.5)
	end

	local ui = form()

	local lists = inv:get_lists()
	for _, l in pairs(lists) do scan_inventory(ui, l) end

	minetest.show_formspec(clicker:get_player_name(), "mymod:inventory_listing_rdonly", ui:render())
end
```

the gracefully-degrading way would be to add a "readonly style" feature flag the mod can check for, and then either send the static inventory list or a readonly `list[]` to a user depending on the value of that flag. temp inventories merely for UI purposes would be an absolute horrorshow, i hope nobody ever does this
